### PR TITLE
App Architecture Improvements

### PR DESCRIPTION
**Pull request by ngmiller**
## What

These changes modify the overall architecture of the front-end. Angular is still the main framework, but utilizing [flux-angular](https://github.com/christianalfoni/flux-angular) allows us to adapt the uni-directional dataflow paradigm put forth by Facebook/React and others.

The build semantics have also been updated slightly. (see README)

## Why

After learning about Facebook's [flux](https://github.com/facebook/flux) architecture, I realized that our setup isn't very flexible. Considering that we may be bringing in and displaying lots of instance or container specific data (i.e. app logs), we'll need an architecture that responds to streaming data and has predictable flows of data and state change.

## How

The overall design pattern is a flavor of MVC, but with an enforced direction of data flow. All data entering the system (i.e. from an API call) is **dispatched** as a **action**. Individual **models** (which flux/flux-angular calls **stores**) may choose to listen to these actions and update their state accordingly. After the state has been updated, a change event is triggered by the model that can be watched by **controllers**, which then make use of the updated data.

Looking at Facebook's documentation on flux gives a pretty good overview of how the architecture is structured, but, in our case, we are not using React and are integrated with Angular.

lighthouse-client-arch-flux-v1

## Happy dance

* Predictable data flow and state updates.
* Can still use two-way data binding on the controller/view interface.
* Minimal integration friction. I was able to iteratively migrate functionality without breaking any other. So, flux-angular and angular coexist peacefully and alongside each other.
* In theory, the entire application state can be saved by serializing the state present within all the active models. One could then reload these models at a future date.
* Introducing bi-directional data flow with the server (as in WebSockets) is trivial. Simply open the connection and push incoming data into the app as a new dispatched action.

## Sad dance

* Currently cannot unit test on Travis without some additional setup. PhantomJS has issues with the JS `bind()` (used by flux-angular) function, but will be updated in the 2.0 release. Until then, unit tests will still work locally (via Chrome) and we can look into [es5-shim](https://github.com/es-shims/es5-shim) for temporary compatability.

## Notes

* In an effort to keep dependencies down, I went ahead and removed restangular since I brought in flux-angular. Angular's provided $http service is certainly good enough for what we need.
* The containerService has been cut down to only a few API calls. It's probably better to just add them as use cases come up.

### CODE REVIEW

**Comment by ngmiller** [comment]
@lighthouse/owners Ready for review.

File: README.md [hunk 1]
```diff
@@ -10,21 +10,31 @@ UI assets and code for Lighthouse.
...
+* `gulp watch` optional, rebuilds on source changes
 * `gulp clean` removes built assets from the server static dir (see Gulpfile.js)
-* `gulp prod build` will do the same as `dev`, but with minified js
-* fire up a webserver from `lighthouse/static` and navigate to `localhost` at the port specified by the server
 
-New angular modules should follow the same pattern as `instances`. Generally speaking, each major piece of functionality should be pulled out into it's own module and given a unique name to link into the root module. (see `app/js/app.js`)
+Options:
+
+* `--prod` build minfied source to `../lighthouse/static`
+* `--gopath` build to `$GOPATH/src/github.com/lighthouse/lighthouse/static/`
```
Comment (ngmiller) [inline README.md]: @theMagicalKarp These are the build updates I mentioned last week. If you're just hacking on the backend and want to build to your go path, `gulp build --gopath` is all you need. And, optionally, `gulp build --gopath --prod`

File: app/js/auth/authService.js [hunk 2]
```diff
@@ -1,106 +1,64 @@
...
+        if (!authModel.isLoggedIn()) {
+            $http.post(request, auth).then(
                 // success
-                function (response) {
-                    var user = {
-                        email: auth.Email
-                    };
-
-                    $rootScope.user = user;
-                    deferred.resolve(user);
```
Comment (ngmiller) [inline app/js/auth/authService.js]: Note: No more nested promises!

**Comment by robertsheehy-wf** [comment]
+1

File: app/js/app.js [hunk 3]
```diff
@@ -25,13 +27,46 @@ var app = angular.module('lighthouse.app', [
...
 
-appConfig.$inject = ['RestangularProvider', '$locationProvider'];
-app.config(appConfig);
+// Initialization
+function appInit($location, flux) {
+    flux.createStore('appStore', {
+        // State
+        route: '',
+
+        // Event hanlders
```
Comment (zach-taylor) [inline app/js/app.js]: hanlders -> handlers

**Comment by zach-taylor** [comment]
LGTM +1

**Pull request merged**